lambda_functions/va_profile/va_profile_opt_in_out_lambda.py

```diff
@@ -0,0 +1,95 @@
+# Tutorial: Configuring a Lambda function to access Amazon RDS in an Amazon VPC
+#   https://docs.aws.amazon.com/lambda/latest/dg/services-rds-tutorial.html
+# https://www.psycopg.org/docs/usage.html
+
+import logging
+import os
+import psycopg2
+import sys
+
+OPT_IN_OUT_QUERY = """SELECT va_profile_opt_in_out(%s, %s, %s, %s, %s);"""
+NOTIFICATION_API_DB_URI = os.getenv("notification_api_db_uri")
+
+if NOTIFICATION_API_DB_URI is None:
+    logging.error("The database URI is not set.")
+    sys.exit("Couldn't connect to the database.")
+
+# https://www.psycopg.org/docs/module.html#exceptions
+try:
+    connection = psycopg2.connect(NOTIFICATION_API_DB_URI)
+except psycopg2.Warning as e:
+    logging.warning(e)
+except psycopg2.Error as e:
+    logging.exception(e)
+
+    # https://www.psycopg.org/docs/errorcodes.html
+    sys.exit(e.pgcode)
```

Comment:
Question about flow. Does it run this once or is this per-call. I suspect it's ran once (when the resource is loaded). Otherwise they'd have to import the modules for every call. What happens if we lose the connection? Also, we'll see a `NameError` if it does not create the connection, which is caught by the generic Exception. What kind of error would we see if the connection was dropped. How long can the connection be left open? Is `sys.exit` valid?

Reply:
I followed the example of app.py in the tutorial linked at the top of the code (line 2): `Executing pymysql.connect() outside of the handler allows your function to re-use the database connection for better performance.`.  I think the significance is that if multiple lambdas are running concurrently, the runtime will only import once, so it doesn't need duplicate connections.  (i.e. what you wrote)  This is also where I got using `sys.exit` in this manner.  If a db connection fails, the module exits, and no lambda handler will run.

I don't see `NameError` in the docs (line 17), but `psycopg2.Error` (line 22) is the base error.  If the connection becomes invalid during execution of the handler, I would expect line 74 to raise an exception that will be caught at line 82.

I'm not aware of any limit on how long a connection can persist.  If a limit exists, I would think that is a database configuration.

Reply:
Ah, I see how they did it, thank you for pointing it out.

The `NameError` would have come from `connection` not being set but a lambda trying to run (line 74). Somewhat related, if we are connected then lose that connection for whatever reason it makes no attempt to reconnect. We're just down until someone restarts the resource. For a more robust system we should handle the case where the connection can be successfully reconnected.